src/components/SVGViewer/pinchzoom.ts

```diff
@@ -0,0 +1,853 @@
+  }
+
+  /**
+   * Scales the zoom factor relative to current state
+   * @param scale
+   * @return the actual scale (can differ because of max min zoom factor)
+   */
+  public scaleZoomFactor(scale: number) {
+    const originalZoomFactor = this.zoomFactor;
+    this.zoomFactor *= scale;
+    this.zoomFactor = Math.min(
+      this.options.maxZoom,
+      Math.max(this.zoomFactor, this.options.minZoom)
+    );
+    return this.zoomFactor / originalZoomFactor;
+  }
+
+  /**
+   * Determine if the image is in a draggable state
+   *
+   * When the image can be dragged, the drag event is acted upon and cancelled.
+   * When not draggable, the drag event bubbles through this component.
+   *
+   * @return {Boolean}
+   */
+  public canDrag() {
+    return (
+      this.options.draggableUnzoomed || !this.isCloseTo(this.zoomFactor, 1)
+    );
+  }
+
+  /**
+   * Drags the element
+   * @param center
+   * @param lastCenter
+   */
+  public drag(center: Coordinates, lastCenter: Coordinates | null) {
+    if (lastCenter) {
+      if (this.options.lockDragAxis) {
+        // lock scroll to position that was changed the most
+        if (
+          Math.abs(center.x - lastCenter.x) > Math.abs(center.y - lastCenter.y)
+        ) {
+          this.addOffset({
+            x: -(center.x - lastCenter.x),
+            y: 0,
+          });
+        } else {
+          this.addOffset({
+            y: -(center.y - lastCenter.y),
+            x: 0,
+          });
+        }
+      } else {
+        this.addOffset({
+          y: -(center.y - lastCenter.y),
+          x: -(center.x - lastCenter.x),
+        });
+      }
+      if (typeof this.options.onDragUpdate == 'function') {
+        this.options.onDragUpdate(this, event);
+      }
+    }
+  }
+
+  /**
+   * Calculates the touch center of multiple touches
+   * @param touches
+   * @return {Object}
+   */
+  public getTouchCenter(touches: Coordinates[]) {
+    return this.getVectorAvg(touches);
+  }
+
+  /**
+   * Calculates the average of multiple vectors (x, y values)
+   */
+  public getVectorAvg(vectors: Coordinates[]) {
+    const sum = (a: number, b: number) => {
+      return a + b;
+    };
+    return {
+      x: vectors.map(v => v.x).reduce(sum) / vectors.length,
+      y: vectors.map(v => v.y).reduce(sum) / vectors.length,
+    };
+  }
+
+  /**
+   * Adds an offset
+   * @param offset the offset to add
+   * @return return true when the offset change was accepted
+   */
+  public addOffset(offset: Coordinates) {
+    this.offset = {
+      x: this.offset.x + offset.x,
+      y: this.offset.y + offset.y,
+    };
+  }
+
+  public sanitize() {
+    if (this.zoomFactor < this.options.zoomOutFactor) {
+      this.zoomOutAnimation();
+    } else if (this.isInsaneOffset(this.offset)) {
+      this.sanitizeOffsetAnimation();
+    }
+  }
+
+  /**
+   * Checks if the offset is ok with the current zoom factor
+   * @param offset
+   * @return {Boolean}
+   */
+  public isInsaneOffset(offset: Coordinates) {
```

Comment:
I don't understand the name of this function lol

Reply:
ahah, well, at least it has jsdoc, so basically checks if offset is a weird value that will cause to show nothing or partly nothing. 

I'm not sure if all that sanitisation is needed at all because I see something being done on svgviewer side as well. But again, just leaving it here for now.